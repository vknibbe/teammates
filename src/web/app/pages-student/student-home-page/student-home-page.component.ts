import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { finalize } from 'rxjs/operators';

import { HttpRequestService } from '../../../services/http-request.service';
import { LoadingBarService } from '../../../services/loading-bar.service';
import { StatusMessageService } from '../../../services/status-message.service';
import { ErrorMessageOutput } from '../../error-message-output';

interface SessionInfoMap {
  endTime: string;
  isOpened: boolean;
  isWaitingToOpen: boolean;
  isPublished: boolean;
  isSubmitted: boolean;
}

interface FeedbackSessionAttributes {
  feedbackSessionName: string;
  courseId: string;
}

interface FeedbackSessionDetailsBundle {
  feedbackSession: FeedbackSessionAttributes;
}

interface StudentCourseAttributes {
  id: string;
  name: string;
}

interface StudentCourse {
  course: StudentCourseAttributes;
  feedbackSessions: FeedbackSessionDetailsBundle[];
}

interface StudentCourses {
  recentlyJoinedCourseId: string;
  hasEventualConsistencyMsg: boolean;
  courses: StudentCourse[];
  sessionsInfoMap: { [key: string]: SessionInfoMap };
}

/**
 * Student home page.
 */
@Component({
  selector: 'tm-student-home-page',
  templateUrl: './student-home-page.component.html',
  styleUrls: ['./student-home-page.component.scss'],
})
export class StudentHomePageComponent implements OnInit {

  // Tooltip messages
  studentFeedbackSessionStatusPublished: string =
      'The responses for the session have been published and can now be viewed.';
  studentFeedbackSessionStatusNotPublished: string =
      'The responses for the session have not yet been published and cannot be viewed.';
  studentFeedbackSessionStatusAwaiting: string =
      'The session is not open for submission at this time. It is expected to open later.';
  studentFeedbackSessionStatusPending: string = 'The feedback session is yet to be completed by you.';
  studentFeedbackSessionStatusSubmitted: string = 'You have submitted your feedback for this session.';
  studentFeedbackSessionStatusClosed: string = ' The session is now closed for submissions.';

  recentlyJoinedCourseId?: string = '';
  hasEventualConsistencyMsg: boolean = false;
  courses: StudentCourse[] = [];
  sessionsInfoMap: { [key: string]: SessionInfoMap } = {};

  constructor(private route: ActivatedRoute, private httpRequestService: HttpRequestService,
              private statusMessageService: StatusMessageService, private loadingBarService: LoadingBarService) { }

  ngOnInit(): void {
    this.route.queryParams.subscribe((queryParams: any) => {
      this.getStudentCourses(queryParams.persistencecourse);
    });
  }

  /**
   * Gets the courses and feedback sessions involving the student.
   */
  getStudentCourses(persistencecourse: string): void {
    this.loadingBarService.showLoadingBar();
    const paramMap: { [key: string]: string } = { persistencecourse };
    this.httpRequestService.get('/student/courses', paramMap)
        .pipe(finalize(() => this.loadingBarService.hideLoadingBar()))
        .subscribe((resp: StudentCourses) => {
          this.recentlyJoinedCourseId = resp.recentlyJoinedCourseId;
          this.hasEventualConsistencyMsg = resp.hasEventualConsistencyMsg;
          this.courses = resp.courses;
          this.sessionsInfoMap = resp.sessionsInfoMap;

          if (this.hasEventualConsistencyMsg) {
            this.statusMessageService.showWarningMessage(
                'You have successfully joined the course ' + `${this.recentlyJoinedCourseId}` + '. '
                + 'Updating of the course data on our servers is currently in progress '
                + 'and will be completed in a few minutes. '
                + 'Please refresh this page in a few minutes to see the course ' + `${this.recentlyJoinedCourseId}`
                + ' in the list below.');
          }

        }, (resp: ErrorMessageOutput) => {
          this.statusMessageService.showErrorMessage(resp.error.message);
        });
  }

  /**
   * Gets the tooltip message for the submission status.
   */
  getSubmissionStatusTooltip(sessionInfoMap: SessionInfoMap): string {
    let msg: string = '';

    if (sessionInfoMap.isWaitingToOpen) {
      msg += this.studentFeedbackSessionStatusAwaiting;
    } else if (sessionInfoMap.isSubmitted) {
      msg += this.studentFeedbackSessionStatusSubmitted;
    } else {
      msg += this.studentFeedbackSessionStatusPending;
    }
    if (!sessionInfoMap.isOpened && !sessionInfoMap.isWaitingToOpen) {
      msg += this.studentFeedbackSessionStatusClosed;
    }
    return msg;
  }

  /**
   * Gets the tooltip message for the response status.
   */
  getResponseStatusTooltip(isPublished: boolean): string {
    if (isPublished) {
      return this.studentFeedbackSessionStatusPublished;
    }
    return this.studentFeedbackSessionStatusNotPublished;
  }
}
