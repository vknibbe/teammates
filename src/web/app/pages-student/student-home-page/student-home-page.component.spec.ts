import { HttpClientTestingModule } from '@angular/common/http/testing';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
import { StudentHomePageComponent } from './student-home-page.component';

import { MatSnackBarModule } from '@angular/material';
import { ResponseStatusPipe } from '../../pipes/session-response-status.pipe';
import { SubmissionStatusPipe } from '../../pipes/session-submission-status.pipe';

describe('StudentHomePageComponent', () => {
  let component: StudentHomePageComponent;
  let fixture: ComponentFixture<StudentHomePageComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        StudentHomePageComponent,
        ResponseStatusPipe,
        SubmissionStatusPipe,
      ],
      imports: [
        HttpClientTestingModule,
        NgbModule,
        RouterTestingModule,
        MatSnackBarModule,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(StudentHomePageComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with no feedback sessions', () => {
    const studentCourse: any = {
      course: {
        id: 'CS3281',
        name: 'Thematic Systems',
      },
      feedbackSessions: [],
    };
    component.courses = [studentCourse];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with all feedback sessions over 2 courses', () => {
    const studentCourseA: any = {
      course: {
        id: 'CS2103',
        name: 'Software Engineering',
      },
      feedbackSessions: [
        {
          feedbackSession: {
            feedbackSessionName: 'First Session',
            courseId: 'CS2103',
          },
        },
        {
          feedbackSession: {
            feedbackSessionName: 'Second Session',
            courseId: 'CS2103',
          },
        },
      ],
    };

    const studentCourseB: any = {
      course: {
        id: 'CS2102',
        name: 'Databases',
      },
      feedbackSessions: [
        {
          feedbackSession: {
            feedbackSessionName: 'Third Session',
            courseId: 'CS2102',
          },
        },
        {
          feedbackSession: {
            feedbackSessionName: 'Fourth Session',
            courseId: 'CS2102',
          },
        },
      ],
    };

    const publishedSessionInfoMap: any = {
      endTime: '1200',
      isOpened: true,
      isWaitingToOpen: true,
      isPublished: true,
      isSubmitted: true,
    };

    const unpublishedSessionInfoMap: any = {
      endTime: '1200',
      isOpened: true,
      isWaitingToOpen: false,
      isPublished: false,
      isSubmitted: false,
    };

    const submittedSessionInfoMap: any = {
      endTime: '1200',
      isOpened: true,
      isWaitingToOpen: false,
      isPublished: false,
      isSubmitted: true,
    };

    const concludedSessionInfoMap: any = {
      endTime: '1200',
      isOpened: false,
      isWaitingToOpen: false,
      isPublished: false,
      isSubmitted: true,
    };

    component.courses = [studentCourseA, studentCourseB];
    component.sessionsInfoMap = {
      'CS2103%First Session': publishedSessionInfoMap,
      'CS2103%Second Session': unpublishedSessionInfoMap,
      'CS2102%Third Session': submittedSessionInfoMap,
      'CS2102%Fourth Session': concludedSessionInfoMap,
    };
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
