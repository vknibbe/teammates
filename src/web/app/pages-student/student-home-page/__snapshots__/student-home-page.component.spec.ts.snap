// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudentHomePageComponent should snap with all feedback sessions over 2 courses 1`] = `
<tm-student-home-page
  courses={[Function Array]}
  hasEventualConsistencyMsg="false"
  httpRequestService={[Function HttpRequestService]}
  loadingBarService={[Function LoadingBarService]}
  recentlyJoinedCourseId=""
  route={[Function ActivatedRoute]}
  sessionsInfoMap={[Function Object]}
  statusMessageService={[Function StatusMessageService]}
  studentFeedbackSessionStatusAwaiting={[Function String]}
  studentFeedbackSessionStatusClosed={[Function String]}
  studentFeedbackSessionStatusNotPublished={[Function String]}
  studentFeedbackSessionStatusPending={[Function String]}
  studentFeedbackSessionStatusPublished={[Function String]}
  studentFeedbackSessionStatusSubmitted={[Function String]}
>
  <div
    class="card bg-light top-padded"
  >
    
    <div
      class="card-header bg-primary text-white"
    >
      <strong>
        [CS2103]: Software Engineering
      </strong>
      <span
        class="float-sm-right"
      >
        <a
          class="btn btn-light btn-sm button"
          href="/course?courseid=CS2103"
          ng-reflect-ngb-tooltip="View and edit information rega"
          ng-reflect-query-params="[object Object]"
          ng-reflect-router-link="../course"
          ngbtooltip="View and edit information regarding your team"
          routerlink="../course"
        >
           View team 
        </a>
      </span>
    </div>
    <div
      class="table-responsive"
    >
      
      <table
        class="table table-striped table-bordered margin-0"
      >
        <thead>
          <tr>
            <th>
              Session Name
            </th>
            <th>
              Deadline
            </th>
            <th>
              Submissions
            </th>
            <th>
              Responses
            </th>
            <th>
              Action(s)
            </th>
          </tr>
        </thead>
        <tbody>
          
          <tr>
            
            
            <td>
              First Session
            </td>
            <td>
              1200
            </td>
            <td>
              <a
                ng-reflect-ngb-tooltip="The session is not open for su"
                style="border-bottom: 1px dashed black;"
              >
                 Submitted 
              </a>
            </td>
            <td>
              <a
                ng-reflect-ngb-tooltip="The responses for the session "
                style="border-bottom: 1px dashed black;"
              >
                 Published 
              </a>
            </td>
            <td>
              
              <a
                class="btn btn-light btn-sm btn-left button"
                href="/sessions/result?courseid=CS2103&fsname=First%20Session"
                ng-reflect-ngb-tooltip="View the submitted responses f"
                ng-reflect-query-params="[object Object]"
                ng-reflect-router-link="../sessions/result"
                ngbtooltip="View the submitted responses for this feedback session"
                routerlink="../sessions/result"
              >
                 View Responses 
              </a>
              
              
              
              <a
                class="btn btn-light btn-sm disabled button"
              >
                 Start Submission 
              </a>
              
              
              <a
                class="btn btn-light btn-sm button"
                href="/sessions/submission?courseid=CS2103&fsname=First%20Session"
                ng-reflect-ngb-tooltip="Edit submitted feedback"
                ng-reflect-query-params="[object Object]"
                ng-reflect-router-link="../sessions/submission"
                ngbtooltip="Edit submitted feedback"
                routerlink="../sessions/submission"
              >
                 Edit Submission 
              </a>
              
              
            </td>
          </tr>
          <tr>
            
            
            <td>
              Second Session
            </td>
            <td>
              1200
            </td>
            <td>
              <a
                ng-reflect-ngb-tooltip="The feedback session is yet to"
                style="border-bottom: 1px dashed black;"
              >
                 Pending 
              </a>
            </td>
            <td>
              <a
                ng-reflect-ngb-tooltip="The responses for the session "
                style="border-bottom: 1px dashed black;"
              >
                 Not Published 
              </a>
            </td>
            <td>
              
              
              <a
                class="btn btn-light btn-sm btn-left disabled button"
                type="button"
              >
                 View Responses 
              </a>
              
              
              <a
                class="btn btn-light btn-sm button"
                href="/sessions/submission?courseid=CS2103&fsname=Second%20Session"
                ng-reflect-ngb-tooltip="Start submitting feedback"
                ng-reflect-query-params="[object Object]"
                ng-reflect-router-link="../sessions/submission"
                ngbtooltip="Start submitting feedback"
                routerlink="../sessions/submission"
              >
                 Start Submission 
              </a>
              
              
              
            </td>
          </tr>
        </tbody>
      </table>
      
    </div>
  </div><div
    class="card bg-light top-padded"
  >
    
    <div
      class="card-header bg-primary text-white"
    >
      <strong>
        [CS2102]: Databases
      </strong>
      <span
        class="float-sm-right"
      >
        <a
          class="btn btn-light btn-sm button"
          href="/course?courseid=CS2102"
          ng-reflect-ngb-tooltip="View and edit information rega"
          ng-reflect-query-params="[object Object]"
          ng-reflect-router-link="../course"
          ngbtooltip="View and edit information regarding your team"
          routerlink="../course"
        >
           View team 
        </a>
      </span>
    </div>
    <div
      class="table-responsive"
    >
      
      <table
        class="table table-striped table-bordered margin-0"
      >
        <thead>
          <tr>
            <th>
              Session Name
            </th>
            <th>
              Deadline
            </th>
            <th>
              Submissions
            </th>
            <th>
              Responses
            </th>
            <th>
              Action(s)
            </th>
          </tr>
        </thead>
        <tbody>
          
          <tr>
            
            
            <td>
              Third Session
            </td>
            <td>
              1200
            </td>
            <td>
              <a
                ng-reflect-ngb-tooltip="You have submitted your feedba"
                style="border-bottom: 1px dashed black;"
              >
                 Submitted 
              </a>
            </td>
            <td>
              <a
                ng-reflect-ngb-tooltip="The responses for the session "
                style="border-bottom: 1px dashed black;"
              >
                 Not Published 
              </a>
            </td>
            <td>
              
              
              <a
                class="btn btn-light btn-sm btn-left disabled button"
                type="button"
              >
                 View Responses 
              </a>
              
              
              
              <a
                class="btn btn-light btn-sm button"
                href="/sessions/submission?courseid=CS2102&fsname=Third%20Session"
                ng-reflect-ngb-tooltip="Edit submitted feedback"
                ng-reflect-query-params="[object Object]"
                ng-reflect-router-link="../sessions/submission"
                ngbtooltip="Edit submitted feedback"
                routerlink="../sessions/submission"
              >
                 Edit Submission 
              </a>
              
              
            </td>
          </tr>
          <tr>
            
            
            <td>
              Fourth Session
            </td>
            <td>
              1200
            </td>
            <td>
              <a
                ng-reflect-ngb-tooltip="You have submitted your feedba"
                style="border-bottom: 1px dashed black;"
              >
                 Closed 
              </a>
            </td>
            <td>
              <a
                ng-reflect-ngb-tooltip="The responses for the session "
                style="border-bottom: 1px dashed black;"
              >
                 Not Published 
              </a>
            </td>
            <td>
              
              
              <a
                class="btn btn-light btn-sm btn-left disabled button"
                type="button"
              >
                 View Responses 
              </a>
              
              
              
              
              <a
                class="btn btn-light btn-sm button"
                href="/sessions/submission?courseid=CS2102&fsname=Fourth%20Session"
                ng-reflect-ngb-tooltip="View submitted feedback"
                ng-reflect-query-params="[object Object]"
                ng-reflect-router-link="../sessions/submission"
                ngbtooltip="View submitted feedback"
                routerlink="../sessions/submission"
              >
                 View Submission 
              </a>
              
            </td>
          </tr>
        </tbody>
      </table>
      
    </div>
  </div>
</tm-student-home-page>
`;

exports[`StudentHomePageComponent should snap with default fields 1`] = `
<tm-student-home-page
  courses={[Function Array]}
  hasEventualConsistencyMsg="false"
  httpRequestService={[Function HttpRequestService]}
  loadingBarService={[Function LoadingBarService]}
  recentlyJoinedCourseId=""
  route={[Function ActivatedRoute]}
  sessionsInfoMap={[Function Object]}
  statusMessageService={[Function StatusMessageService]}
  studentFeedbackSessionStatusAwaiting={[Function String]}
  studentFeedbackSessionStatusClosed={[Function String]}
  studentFeedbackSessionStatusNotPublished={[Function String]}
  studentFeedbackSessionStatusPending={[Function String]}
  studentFeedbackSessionStatusPublished={[Function String]}
  studentFeedbackSessionStatusSubmitted={[Function String]}
>
  
</tm-student-home-page>
`;

exports[`StudentHomePageComponent should snap with no feedback sessions 1`] = `
<tm-student-home-page
  courses={[Function Array]}
  hasEventualConsistencyMsg="false"
  httpRequestService={[Function HttpRequestService]}
  loadingBarService={[Function LoadingBarService]}
  recentlyJoinedCourseId=""
  route={[Function ActivatedRoute]}
  sessionsInfoMap={[Function Object]}
  statusMessageService={[Function StatusMessageService]}
  studentFeedbackSessionStatusAwaiting={[Function String]}
  studentFeedbackSessionStatusClosed={[Function String]}
  studentFeedbackSessionStatusNotPublished={[Function String]}
  studentFeedbackSessionStatusPending={[Function String]}
  studentFeedbackSessionStatusPublished={[Function String]}
  studentFeedbackSessionStatusSubmitted={[Function String]}
>
  <div
    class="card bg-light top-padded"
  >
    
    <div
      class="card-header bg-primary text-white"
    >
      <strong>
        [CS3281]: Thematic Systems
      </strong>
      <span
        class="float-sm-right"
      >
        <a
          class="btn btn-light btn-sm button"
          href="/course?courseid=CS3281"
          ng-reflect-ngb-tooltip="View and edit information rega"
          ng-reflect-query-params="[object Object]"
          ng-reflect-router-link="../course"
          ngbtooltip="View and edit information regarding your team"
          routerlink="../course"
        >
           View team 
        </a>
      </span>
    </div>
    <div
      class="table-responsive"
    >
      
      
      <table
        class="table table-striped table-bordered margin-0"
      >
        <tbody>
          <tr>
            <th
              class="align-center white"
            >
               Currently, there are no open evaluation/feedback sessions in this course. When a session is open for submission you will be notified. 
            </th>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</tm-student-home-page>
`;
